includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,325 @@
+	 *
+	 * If it's not valid AMP, it adds a query arg to the redirect URL.
+	 * This will cause an error message to appear above the 'Classic' editor.
+	 *
+	 * @param integer $post_id The ID of the updated post.
+	 * @param WP_Post $post The updated post.
+	 * @return void.
+	 */
+	public static function validate_content( $post_id, $post ) {
+		if ( ! self::is_authorized() ) {
+			return;
+		}
+		$filtered_content = apply_filters( 'the_content', $post->post_content );
+		$response         = self::get_response( $filtered_content );
+		if ( isset( $response[ self::ERROR_KEY ] ) && ( true === $response[ self::ERROR_KEY ] ) ) {
+			add_filter( 'redirect_post_location', array( __CLASS__, 'error_message' ) );
+		}
+	}
+
+	/**
+	 * Whether the current user is authorized.
+	 *
+	 * This checks that the user has a certain capability and the nonce is valid.
+	 * It will only return true when updating the post on:
+	 * wp-admin/post.php
+	 * Avoids using check_admin_referer().
+	 * This function might be called in different places,
+	 * and it can't cause it to die() if the nonce is invalid.
+	 *
+	 * @return boolean $is_valid True if the nonce is valid.
+	 */
+	public static function is_authorized() {
+		$nonce = isset( $_REQUEST['_wpnonce'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['_wpnonce'] ) ) : ''; // WPCS: CSRF ok.
+		return ( self::has_cap() && ( false !== wp_verify_nonce( $nonce, 'update-post_' . get_the_ID() ) ) );
+	}
+
+	/**
+	 * Output AMP validation data in the response header of a frontend GET request.
+	 *
+	 * This must be called before the document output begins.
+	 * Because the document is buffered,
+	 * The sanitizers run after the 'send_headers' action.
+	 * So it's not possible to call this function on that hook.
+	 *
+	 * @return void.
+	 */
+	public static function add_header() {
+		if ( self::is_authorized() ) {
+			header( sprintf( 'AMP-Validation-Error: %s', wp_json_encode( self::get_response() ) ) );
+		}
+	}
+
+	/**
+	 * Adds an error message to the URL if it's not valid AMP.
+	 *
+	 * When redirecting after saving a post, the content was validated for AMP compliance.
+	 * If it wasn't valid AMP, this will add a query arg to the URL.
+	 * And an error message will display on /wp-admin/post.php.
+	 *
+	 * @param string $url The URL of the redirect.
+	 * @return string $url The filtered URL, including the AMP error message query var.
+	 */
+	public static function error_message( $url ) {
+		return add_query_arg(
+			self::ERROR_QUERY_KEY,
+			self::ERROR_QUERY_VALUE,
+			$url
+		);
+	}
+
+	/**
+	 * Displays an error message on /wp-admin/post.php if the saved content is not valid AMP.
+	 *
+	 * Use $_GET, as get_query_var won't return the value.
+	 * This displays at the top of the 'Classic' editor.
+	 *
+	 * @return void.
+	 */
+	public static function display_error() {
+		$error = isset( $_GET[ self::ERROR_QUERY_KEY ] ) ? sanitize_text_field( wp_unslash( $_GET[ self::ERROR_QUERY_KEY ] ) ) : ''; // WPCS: CSRF ok.
```

Comment:
For compliance sake, it would be good to check nonce here.

Reply:
Good point, @ThierryA. This [commit](https://github.com/Automattic/amp-wp/commit/583a6bcf549283c167673624ed5be141f3be347a) adds nonce verification, via `check_admin_referer()`.